File: janitor/debian/dist.py
#!/usr/bin/python3
# Copyright (C) 2018 Jelmer Vernooij <jelmer@jelmer.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

from contextlib import ExitStack
import errno
import logging
import os
import sys
import traceback

from ognibuild.session import SessionSetupFailure
from ognibuild.dist import (
    DIST_LOG_FILENAME,
    dist,
    DistNoTarball,
)
from ognibuild import (
    DetailedFailure,
    UnidentifiedError,
)
from ognibuild.buildsystem import (
    NoBuildToolsFound,
)
from ognibuild.logs import (
    DirectoryLogManager,
    NoLogManager,
)


logger = logging.getLogger(__name__)


def report_failure(kind, description, original):
    logging.fatal('%s: %s', kind, description)
    if 'DIST_RESULT' in os.environ:
        import json
        with open(os.environ['DIST_RESULT'], 'w') as f:
            json.dump({
                'result_code': kind,
                'description': description}, f)


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--log-directory', type=str, default=None,
        help='Write logs to files in specified directory rather than standard out')
    parser.add_argument(
        '--schroot', type=str, default=os.environ.get('SCHROOT'),
        help='Schroot to use')
    parser.add_argument(
        '--target-dir', type=str, default='..',
        help='Directory to write to')
    parser.add_argument(
        '--packaging', type=str, default=None,
        help='Location of packaging')
    parser.add_argument(
        '--directory', '-d',
        type=str, default='.',
        help='Path to tree to create dist tarball for')
    parser.add_argument(
        '--require-declared',
        action='store_true',
        help='Fail if declared dependencies can not be installed')
    args = parser.parse_args()

    from ognibuild.session.schroot import SchrootSession

    import breezy.bzr  # noqa: F401
    import breezy.git  # noqa: F401
    from breezy.errors import NotBranchError
    from breezy.workingtree import WorkingTree

    logging.basicConfig(level=logging.INFO, format='%(message)s')

    package = os.environ.get('PACKAGE')
    version = os.environ.get('VERSION')

    with ExitStack() as es:
        subdir = package or "package"

        session = es.enter_context(SchrootSession(args.schroot))
        try:
            try:
                tree = WorkingTree.open(args.directory)
            except NotBranchError:
                export_directory, reldir = session.setup_from_directory(
                    args.directory, subdir)
            else:
                export_directory, reldir = session.setup_from_vcs(
                    tree, include_controldir=True, subdir=subdir
                )
        except OSError as e:
            if e.errno == errno.ENOSPC:
                report_failure(
                    'no-space-on-device', 'No space on device running mkdtemp',
                    e)
                sys.exit(1)
            raise

        if args.packaging:
            (packaging_tree,
             packaging_debian_path) = WorkingTree.open_containing(
                args.packaging)
            from ognibuild.debian import satisfy_build_deps

            try:
                satisfy_build_deps(
                    session, packaging_tree, packaging_debian_path)
            except DetailedFailure as e:
                logging.warning(
                    'Ignoring error installing declared build dependencies '
                    '(%s): %s', e.error.kind, str(e.error))
                if args.require_declared:
                    sys.exit(1)
            except UnidentifiedError as e:
                lines = [line for line in e.lines if line]
                if e.secondary:
                    logging.warning(
                        'Ignoring error installing '
                        'declared build dependencies (%r): %s',
                        e.argv, e.secondary.line)
                    report_failure('dist-command-failed', e.secondary.line, e)
                elif len(lines) == 1:
                    logging.warning(
                        'Ignoring error installing declared '
                        'build dependencies (%r): %s',
                        e.argv, lines[0])
                else:
                    logging.warning(
                        'Ignoring error installing declared '
                        'build dependencies (%r): %r',
                        e.argv, lines)
                if args.require_declared:
                    sys.exit(1)
        else:
            packaging_tree = None
            packaging_debian_path = None

        try:
            target_dir = os.path.abspath(
                os.path.join(args.directory, args.target_dir))

            if args.log_directory:
                log_manager = DirectoryLogManager(
                    os.path.join(args.log_directory, DIST_LOG_FILENAME),
                    mode='redirect')
            else:
                log_manager = NoLogManager()

            try:
                dist(session, export_directory, reldir, target_dir,
                     version=version, log_manager=log_manager)
            except NotImplementedError:
                sys.exit(2)
            except NoBuildToolsFound:
                logger.info(
                    "No build tools found, falling back to simple export.")
                sys.exit(2)
        except UnidentifiedError as e:
            lines = [line for line in e.lines if line]
            if e.secondary:
                report_failure('dist-command-failed', e.secondary.line, e)
            elif len(lines) == 1:
                report_failure('dist-command-failed', lines[0], e)
            else:
                report_failure(
                    'dist-command-failed',
                    "%r failed with unidentified error "
                    "(return code %d)" % (e.argv, e.retcode), e)
            sys.exit(1)
        except SessionSetupFailure as e:
            report_failure('session-setup-failure', str(e), e)
            sys.exit(1)
        except DetailedFailure as e:
            if e.error.is_global:
                error_code = e.error.kind
            else:
                error_code = "dist-" + e.error.kind
            error_description = str(e.error)
            report_failure(error_code, error_description, e)
            sys.exit(1)
        except DistNoTarball as e:
            report_failure('dist-no-tarball', str(e), e)
            sys.exit(1)
        except BaseException as e:
            traceback.print_exc()
            report_failure('dist-exception', str(e), e)
            sys.exit(1)

    sys.exit(0)

File: worker/src/bin/dist.rs
use breezyshim::error::Error as BrzError;
use clap::Parser;
use ognibuild::analyze::AnalyzedError;
use ognibuild::buildsystem::Error;
use ognibuild::dist::dist;
use ognibuild::logs::{DirectoryLogManager, LogManager, LogMode, NoLogManager};
use ognibuild::session::Session;
use std::path::PathBuf;

#[derive(Parser)]
struct Args {
    #[clap(short, long)]
    /// Write logs to files in specified directory rather than standard out
    log_directory: Option<PathBuf>,

    #[cfg(target_os = "linux")]
    #[clap(short, long, env = "SCHROOT")]
    /// Schroot to use
    schroot: Option<String>,

    #[clap(short, long, default_value = "..")]
    /// Directory to write to
    target_dir: PathBuf,

    #[clap(short, long)]
    /// Path to packaging tree
    packaging: Option<PathBuf>,

    #[clap(short, long, default_value = ".")]
    /// Path to tree to create dist tarball for
    directory: PathBuf,

    #[clap(long)]
    /// Fail if declared dependencies can not be installed
    require_declared: bool,

    #[clap(long)]
    apt_update: bool,

    #[clap(long)]
    apt_dist_upgrade: bool,

    #[command(flatten)]
    logging: janitor::logging::LoggingArgs,
}

fn report_failure(kind: &str, description: &str) -> ! {
    log::error!("{}: {}", kind, description);
    if let Ok(path) = std::env::var("DIST_RESULT") {
        let result = serde_json::json!({
            "result_code": kind,
            "description": description,
        });
        std::fs::write(path, serde_json::to_string(&result).unwrap()).unwrap();
    }
    std::process::exit(1);
}

fn main() -> Result<(), i32> {
    let args = Args::parse();

    args.logging.init();

    let package = std::env::var("PACKAGE").ok();
    let version = std::env::var("VERSION").ok();

    let subdir = package.unwrap_or_else(|| "package".to_owned());

    #[cfg(target_os = "linux")]
    let mut session: Box<dyn Session> = if let Some(schroot) = args.schroot {
        Box::new(ognibuild::session::schroot::SchrootSession::new(&schroot, None).unwrap())
    } else {
        Box::new(ognibuild::session::plain::PlainSession::new())
    };

    #[cfg(not(target_os = "linux"))]
    let mut session: Box<dyn Session> = Box::new(ognibuild::session::plain::PlainSession::new());

    #[cfg(feature = "debian")]
    if args.apt_update {
        ognibuild::debian::apt::run_apt(session.as_ref(), vec!["update"], vec![]).unwrap();
    }
    #[cfg(feature = "debian")]
    if args.apt_dist_upgrade {
        ognibuild::debian::apt::run_apt(session.as_ref(), vec!["dist-upgrade"], vec![]).unwrap();
    }

    let project = match breezyshim::workingtree::open(&args.directory) {
        Ok(tree) => session
            .project_from_vcs(&tree, Some(true), Some(&subdir))
            .unwrap(),
        Err(BrzError::NotBranchError(..)) => session
            .project_from_directory(&args.directory, Some(&subdir))
            .unwrap(),
        Err(e) => {
            report_failure("vcs-error", &format!("Error opening working tree: {}", e));
        }
    };

    #[cfg(feature = "debian")]
    let (packaging_tree, packaging_debian_path) = if let Some(packaging) = args.packaging {
        let (packaging_tree, packaging_debian_path) =
            breezyshim::workingtree::open_containing(&packaging).unwrap();

        match ognibuild::debian::satisfy_build_deps(
            session.as_ref(),
            &packaging_tree,
            &packaging_debian_path,
        ) {
            Ok(_) => (Some(packaging_tree), Some(packaging_debian_path)),
            Err(ognibuild::debian::apt::Error::Detailed { error, .. }) => {
                let error = error.unwrap();
                log::warn!(
                    "Ignoring error installing declared build dependencies ({}): {:?}",
                    error.kind(),
                    error
                );
                if args.require_declared {
                    report_failure(error.kind().as_ref(), &error.to_string());
                }
                (None, None)
            }
            Err(ognibuild::debian::apt::Error::Unidentified {
                lines,
                secondary,
                args: argv,
                ..
            }) => {
                if let Some(secondary) = secondary {
                    log::warn!(
                        "Ignoring error installing declared build dependencies ({:?}): {}",
                        argv,
                        secondary.line()
                    );
                    if args.require_declared {
                        report_failure("command-failed", &secondary.line());
                    }
                } else if lines.len() == 1 {
                    log::warn!(
                        "Ignoring error installing declared build dependencies ({:?}): {}",
                        argv,
                        lines[0]
                    );
                } else {
                    log::warn!(
                        "Ignoring error installing declared build dependencies ({:?}): {:?}",
                        argv,
                        lines
                    );
                }
                if args.require_declared {
                    report_failure("command-failed", &lines.join("\n"));
                }
                (None, None)
            }
            Err(ognibuild::debian::apt::Error::Session(
                ognibuild::session::Error::SetupFailure(e, _),
            )) => {
                report_failure("session-setup-failure", &e.to_string());
            }
            Err(ognibuild::debian::apt::Error::Session(ognibuild::session::Error::IoError(e))) => {
                report_failure("session-io-error", &e.to_string());
            }
            Err(ognibuild::debian::apt::Error::Session(
                ognibuild::session::Error::CalledProcessError(e),
            )) => {
                report_failure("session-process-error", &e.to_string());
            }
        }
    } else {
        (None, None)
    };

    let target_dir = args
        .directory
        .join(&args.target_dir)
        .canonicalize()
        .unwrap();

    let mut log_manager: Box<dyn LogManager> = if let Some(log_directory) = args.log_directory {
        Box::new(DirectoryLogManager::new(
            log_directory.join("dist.log"),
            LogMode::Redirect,
        ))
    } else {
        Box::new(NoLogManager)
    };

    match dist(
        session.as_mut(),
        project.external_path(),
        project.internal_path(),
        &target_dir,
        log_manager.as_mut(),
        version.as_deref(),
        !args.logging.debug,
    ) {
        Ok(t) => {
            log::info!("Dist tarball {} created successfully", t.to_str().unwrap());
        }
        Err(Error::Unimplemented) => {
            return Err(2);
        }
        Err(Error::NoBuildSystemDetected) => {
            log::warn!("No build system detected, falling back to simple export.");
            return Err(2);
        }
        Err(Error::Error(AnalyzedError::Unidentified {
            retcode,
            lines,
            secondary,
        })) => {
            if let Some(secondary) = secondary {
                report_failure("command-failed", &secondary.line());
            } else if lines.len() == 1 {
                report_failure("command-failed", &lines[0]);
            } else {
                report_failure("command-failed", &lines.join("\n"));
            }
        }
        Err(Error::Error(AnalyzedError::Detailed { error, .. })) => {
            report_failure(error.kind().as_ref(), &error.to_string());
        }
        Err(Error::Error(AnalyzedError::IoError(e))) => {
            report_failure("io-error", &e.to_string());
        }
        Err(Error::Error(AnalyzedError::MissingCommandError { command })) => {
            report_failure("missing-command", &format!("Missing command: {}", command));
        }
        Err(Error::Other(e)) => {
            report_failure("other", &format!("Unknown error: {}", e));
        }
        Err(Error::DependencyInstallError(e)) => {
            report_failure("dependency-install-error", "Failed to install dependencies");
        }
        Err(Error::IoError(e)) => {
            report_failure("io-error", &e.to_string());
        }
    }

    Ok(())
}
